### PR TITLE
fix typed list equals

### PR DESCRIPTION
**Pull request by esc**
The previous implementation incorrectly assumed that when comparing
equality of lists, we can compare the list elements for order. However,
not all possible types can be compared for order, for example typed
dictionaries can be compared for equality but not for order and so the
previous  implementation failed during typing. This fixes that and the
new equals and not equals methods don't assume that the types can be
compared for order but for quality only.

### CODE REVIEW

**Comment by esc** [comment]
Fixes #4879

File: numba/listobject.py [hunk 1]
```diff
@@ -1038,6 +1038,42 @@ def impl(l, item, start=None, end=None):
...
+    if op == "equals":
+        def impl(this, other):
+            return equals(this, other)
+    elif op == "not_equals":
+        def impl(this, other):
+            return not equals(this, other)
+    return impl
+
+
+@overload(operator.eq)
```
Comment (stuartarchibald) [inline numba/listobject.py]: Wondering if it's worth doing something like:
```python
@overload(operator.truth)
def ol_truth(val):
    if isinstance(val, types.Boolean):
        def impl(val):
            return val
        return impl

def _equals_helper(this, other, OP):
    if not isinstance(this, types.ListType):
        return
    if not isinstance(other, types.ListType):
        raise TypingError("list can only be compared to list")

    def impl(this, other):
        def equals(this, other):
            if len(this) != len(other):
                return False
            for i in range(len(this)):
                if this[i] != other[i]:
                    return False
            else:
                return True
        return OP(equals(this, other))
    return impl


@overload(operator.eq)
def impl_equals(this, other):
    return _equals_helper(this, other, operator.truth)


@overload(operator.ne)
def impl_not_equals(this, other):
    return _equals_helper(this, other, operator.not_)
```
saves on switching on strings.

File: numba/listobject.py [hunk 2]
```diff
@@ -1038,6 +1038,42 @@ def impl(l, item, start=None, end=None):
...
 
 
+@register_jitable
+def equals(this, other):
+    if len(this) != len(other):
+        return False
+    for i in range(len(this)):
+        if this[i] != other[i]:
+            return False
+    else:
```
[the remark this answers is not in the document]
Comment (esc) [reply numba/listobject.py]: no, this is just for style.

File: numba/listobject.py [hunk 3]
```diff
@@ -1038,6 +1038,42 @@ def impl(l, item, start=None, end=None):
...
+        return True
+
+
+def _equals_helper(this, other, op):
+    if not isinstance(this, types.ListType):
+        return
+    if not isinstance(other, types.ListType):
```
[the remark this answers is not in the document]
Comment (esc) [reply numba/listobject.py]: agree, I'll fix it in the compare-helper too.

File: numba/listobject.py (hunk 1 above)
Comment (esc) [reply numba/listobject.py]: Yes! 👀

**Comment by esc** [comment]
@stuartarchibald I believe I have addressed all of your comments and this PR is now ready for the next round of review.

**Comment by esc** [comment]
Thanks for merging.

**Pull request merged**